src/Cake.Common/Tools/OctopusDeploy/OctopusDeployPusher.cs

```diff
@@ -0,0 +1,82 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Cake.Core;
+using Cake.Core.IO;
+using Cake.Core.Tooling;
+
+namespace Cake.Common.Tools.OctopusDeploy
+{
+    /// <summary>
+    /// The Octopus Deploy package push runner
+    /// </summary>
+    public class OctopusDeployPusher : Tool<PushPackageSettings>
+    {
+        private readonly ICakeEnvironment _environment;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OctopusDeployPusher"/> class.
+        /// </summary>
+        /// <param name="fileSystem">The file system.</param>
+        /// <param name="environment">The environment.</param>
+        /// <param name="processRunner">The process runner.</param>
+        /// <param name="tools">The tool locator.</param>
+        public OctopusDeployPusher(
+            IFileSystem fileSystem, 
+            ICakeEnvironment environment, 
+            IProcessRunner processRunner, 
+            IToolLocator tools) : base(fileSystem, environment, processRunner, tools)
+        {
+            _environment = environment;
+        }
+
+        /// <summary>
+        /// Gets the name of the tool.
+        /// </summary>
+        /// <returns>The name of the tool.</returns>
+        protected override string GetToolName()
+        {
+            return "Octo";
+        }
+
+        /// <summary>
+        /// Gets the possible names of the tool executable.
+        /// </summary>
+        /// <returns>The tool executable name.</returns>
+        protected override IEnumerable<string> GetToolExecutableNames()
+        {
+            return new[] { "Octo.exe" };
+        }
+
+        /// <summary>
+        /// Pushes the specified packages to Octopus Deploy internal repository
+        /// </summary>
+        /// <param name="settings">The settings</param>
+        /// <param name="packagePaths">Paths to the packages to be pushed</param>
+        public void PushPackage(PushPackageSettings settings, params FilePath[] packagePaths)
+        {
+            if (packagePaths == null || !packagePaths.Any())
+            {
+                throw new ArgumentNullException("packagePaths");
+            }
+            if (settings == null)
+            {
+                throw new ArgumentNullException("settings");
+            }
+            if (string.IsNullOrEmpty(settings.Server))
+            {
+                throw new ArgumentException("No server specified.", "settings");
```

Comment:
Okay, so just to be clear I will include `server` and `apiKey` in the constructor parameters, but which do we trust: the raw params or the values from the settings object?

Also, just so I am aware, what is the reasoning behind not putting these in a settings class since they are going to be common to all Octopus Deploy aliases (including future ones, if the case arises)?